group-coordinator/src/main/java/org/apache/kafka/coordinator/group/assignor/OptimizedUniformAssignmentBuilder.java

```diff
@@ -0,0 +1,393 @@
+        return new GroupAssignment(newAssignment);
+    }
+
+    /**
+     * Retrieves a set of partitions that were currently assigned to members and will be retained in the new assignment,
+     * by ensuring that the partitions are still relevant based on current topic metadata and subscriptions.
+     * If rack awareness is enabled, it ensures that a partition's rack matches the member's rack.
+     *
+     * <p> For each member, it:
+     * <ul>
+     *     <li> Finds the valid current assignment considering topic subscriptions and metadata.</li>
+     *     <li> If current assignments exist, retains up to the minimum quota of assignments.</li>
+     *     <li> If there are members that should get an extra partition,
+     *          assigns the next partition after the retained ones.</li>
+     *     <li> For members with assignments not exceeding the minimum quota,
+     *          it identifies them as potentially unfilled members and tracks the remaining quota.</li>
+     * </ul>
+     *
+     * @return A set containing all the sticky partitions that have been retained in the new assignment.
+     */
+    private Set<TopicIdPartition> computeAssignedStickyPartitions(Integer minQuota) {
+        Set<TopicIdPartition> allAssignedStickyPartitions = new HashSet<>();
+
+        assignmentSpec.members().forEach((memberId, assignmentMemberSpec) -> {
+            // Remove all the topics that aren't in the subscriptions or the topic metadata anymore.
+            // If rack awareness is enabled, only add partitions if the members rack matches the partitions rack.
+            List<TopicIdPartition> validCurrentAssignment = validCurrentAssignment(
+                memberId,
+                assignmentMemberSpec.assignedPartitions()
+            );
+
+            int currentAssignmentSize = validCurrentAssignment.size();
+            int remaining = minQuota - currentAssignmentSize;
+
+            if (currentAssignmentSize > 0) {
+                int retainedPartitionsCount = min(currentAssignmentSize, minQuota);
+                IntStream.range(0, retainedPartitionsCount).forEach(i -> {
+                    newAssignment.get(memberId)
+                        .targetPartitions()
+                        .computeIfAbsent(validCurrentAssignment.get(i).topicId(), __ -> new HashSet<>())
+                        .add(validCurrentAssignment.get(i).partition());
+                    allAssignedStickyPartitions.add(validCurrentAssignment.get(i));
+                });
+
+                // The extra partition is located at the last index from the previous step.
+                if (remaining < 0 && remainingMembersToGetAnExtraPartition > 0) {
+                    newAssignment.get(memberId)
+                        .targetPartitions()
+                        .computeIfAbsent(validCurrentAssignment.get(retainedPartitionsCount).topicId(), __ -> new HashSet<>())
+                        .add(validCurrentAssignment.get(retainedPartitionsCount).partition());
```

Comment:
nit: This code is duplicated. Should we have an helper method for it?

Reply:
will do